src/source/agc.rs

```diff
@@ -0,0 +1,345 @@
+///
+/// This struct implements an AGC algorithm that dynamically adjusts audio levels
+/// based on both peak and RMS (Root Mean Square) measurements.
+#[derive(Clone, Debug)]
+pub struct AutomaticGainControl<I> {
+    input: I,
+    target_level: f32,
+    absolute_max_gain: f32,
+    current_gain: f32,
+    attack_coeff: f32,
+    release_coeff: f32,
+    min_attack_coeff: f32,
+    peak_level: f32,
+    rms_window: CircularBuffer,
+    is_enabled: Arc<AtomicBool>,
+}
+
+/// A circular buffer for efficient RMS calculation over a sliding window.
+///
+/// This structure allows for constant-time updates and mean calculations,
+/// which is crucial for real-time audio processing.
+#[derive(Clone, Debug)]
+struct CircularBuffer {
+    buffer: [f32; RMS_WINDOW_SIZE],
+    index: usize,
+    sum: f32,
+}
+
+impl CircularBuffer {
+    /// Creates a new CircularBuffer with a fixed size determined at compile time.
+    ///
+    /// The `_size` parameter is ignored as the buffer size is set by `RMS_WINDOW_SIZE`.
+    #[inline]
+    fn new(_size: usize) -> Self {
+        CircularBuffer {
+            buffer: [0.0; RMS_WINDOW_SIZE],
+            index: 0,
+            sum: 0.0,
+        }
+    }
+
+    /// Pushes a new value into the buffer and returns the old value.
+    ///
+    /// This method maintains a running sum for efficient mean calculation.
+    #[inline]
+    fn push(&mut self, value: f32) -> f32 {
+        let old_value = self.buffer[self.index];
+        self.buffer[self.index] = value;
+        self.sum += value - old_value;
+        self.index = (self.index + 1) % self.buffer.len();
+        old_value
+    }
+
+    /// Calculates the mean of all values in the buffer.
+    ///
+    /// This operation is O(1) due to the maintained running sum.
+    #[inline]
+    fn mean(&self) -> f32 {
+        self.sum / self.buffer.len() as f32
+    }
+}
+
+/// Constructs an `AutomaticGainControl` object with specified parameters.
+///
+/// # Arguments
+///
+/// * `input` - The input audio source
+/// * `target_level` - The desired output level
+/// * `attack_time` - Time constant for gain increase
+/// * `release_time` - Time constant for gain decrease
+/// * `absolute_max_gain` - Maximum allowable gain
+#[inline]
+pub fn automatic_gain_control<I>(
+    input: I,
+    target_level: f32,
+    attack_time: f32,
+    release_time: f32,
+    absolute_max_gain: f32,
+) -> AutomaticGainControl<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    let sample_rate = input.sample_rate();
+
+    AutomaticGainControl {
+        input,
+        target_level,
+        absolute_max_gain,
+        current_gain: 1.0,
+        attack_coeff: (-1.0 / (attack_time * sample_rate as f32)).exp(),
+        release_coeff: (-1.0 / (release_time * sample_rate as f32)).exp(),
+        min_attack_coeff: release_time,
+        peak_level: 0.0,
+        rms_window: CircularBuffer::new(RMS_WINDOW_SIZE),
+        is_enabled: Arc::new(AtomicBool::new(true)),
```

Comment:
I see you added 'dynamic' controls in there. The current way we do controls in rodio is through the sink. @PetrGlad had the idea to try and do this with Arcs + Atomics just like you do here.  I have no idea what the performance impact of this is, its something I have been discussing in https://github.com/RustAudio/rodio/issues/619 and which will move to its own issue in the future. 

For now its a bit strange that `enabled` can be changed via a handle (gotten through `get_agc_control`) but the other variables can not. @PetrGlad AGC might be a good candidate for benchmarking dynamic controls (though atomics+arcs) since it has so many variables.

Reply:
> idea to try and do this with Arcs + Atomics just like you do here. I have no idea what the performance impact of this is

Almost negligible especially with atomics it almost is around the same level of performance as an if statement or a conversion to f32 i.e. in µs/micro seconds

but more than welcome to benchmark it somehow

Reply:
well since your experimenting (and having fun) anyways :)

I suspect that performance impact (9.5 vs 7.6 is a significant increase) does not come from the fact that we are using atomics but the Arc's. Where all these controls where in the same struct next to each-other now they are spread over the heap. That makes the optimizer struggle to optimize more and its difficult for the cpu to keep everything in registers/L1 cache.

So maybe we can adress them by using only one Arc over a struct of atomics?

Reply:
On the other hind I kinda wanna merge this without any of the atomic/arc stuff (we need to hammer out a clear API and apply that to everything and this PR is not the place for that). So if you want to experiment be my guest :) but to merge it we'll need it plain and boring again.